### ISSUE TITLE
MongoDB client crashes when using UTF-8 special characters

### ISSUE THREAD

**Issue by danijar** [share/sharedb#428]
We are using ShareDB (sharedb@1.4.1) with the MongoDB adapter (sharedb-mongo@1.0.0-beta.14) to synchronize a text editor (sharedb-codemirror@1.0.0) between multiple collaborators.

When entering and deleting multi-byte UTF-8 characters (e.g. 😀) into the text editor, the ShareDB backend of our server crashes:

```
Error: Delete component '�' does not match deleted text '�'
    at Object.text.apply (/usr/src/app/server/node_modules/ot-json0/lib/text0.js:71:15)
    at Object.json.apply (/usr/src/app/server/node_modules/ot-json0/lib/json0.js:171:33)
    at Object.exports.applyOps (/usr/src/app/server/node_modules/sharedb/lib/ot.js:185:28)
    at Backend._buildSnapshotFromOps (/usr/src/app/server/node_modules/sharedb/lib/backend.js:826:18)
    at /usr/src/app/server/node_modules/sharedb/lib/backend.js:754:15
    at /usr/src/app/server/node_modules/sharedb-mongo/index.js:403:7
    at /usr/src/app/server/node_modules/mongodb/lib/utils.js:674:5
    at handleCallback (/usr/src/app/server/node_modules/mongodb/lib/utils.js:102:55)
    at /usr/src/app/server/node_modules/mongodb/lib/cursor.js:838:66
    at /usr/src/app/server/node_modules/mongodb/lib/utils.js:674:5
```

Specifically, when inserting an emoji 😀 and pressing backspace, the emoji turns into �. When pressing backspace again, both the � and the preceding character are removed. After that, inserting any other character crashes the server.

**Comment by alecgibson** [comment]
I _think_ this is a limitation with the `rich-text` type, and which is why there's [`text-unicode`][1]. You can try installing and [subtyping][2] your documents with that type.

That being said, I'd also be a bit suspicious of your front-end code. Most browsers should already treat emoji in a sensible manner (ie attempt to remove all relevant unicode points). This smells of something trying to manually delete a character byte-wise, which isn't strictly correct (as you can see).

[1]: https://github.com/ottypes/text-unicode
[2]: https://github.com/ottypes/json0#subtype-operations

**Comment by danijar** [comment]
Thanks for the pointer. We've fixed the frontend now but would still like to safeguard the server from crashing when receiving problematic transactions.

The subtyping link you shared makes it seem like the type annotations are added to the operations, not the documents so an attacker could still send operations to crash the server?

Ideally, ShareDB would robust against invalid operations and reject them with an error message rather than crashing the server.

**Comment by alecgibson** [comment]
I fixed this particular issue in https://github.com/share/sharedb/pull/399 which was released in [`v1.4.2`](https://github.com/share/sharedb/releases/tag/v1.4.2). Please upgrade your version of `sharedb`.

**Comment by danijar** [comment]
Awesome, thanks!
